### ISSUE TITLE
Support Sidekiq required feature set

### ISSUE THREAD

**Issue by casperisfine** [redis-rb/redis-client#6]
As discussed in https://github.com/redis/redis-rb/issues/1070#issuecomment-1074250310, I'm trying to convert Sidekiq to use `redis-client`, so far the missing features I found:

  - [x] `url:` connection param. 
  - [x] `id` connection param (`CLIENT SETNAME`)
  - [x] helper for `SCAN` family of commands
  - [x] ~~`Redis::Namespace` equivalent (this one is debatable, could potentially be implemented as some kind of middleware that has `COMMAND` knowledge)~~
  - [x] `reconnect_attempts`, Sidekiq sets it to `1`. However if we are to implement this, we should have `COMMAND` knowledge to only retry `readonly` tagged commands.
  - [X] Empty pipelines. Sidekiq test suite allowed to noticed empty pipeline were crashing.
  - [x] The sidekiq test suite has some Redis sentinel configuration, we should support that as well (#8).

**Comment by casperisfine** [comment]
test

**Comment by mperham** [comment]
I'm happy to adjust or remove some feature usage. I tried to remove `reconnect_attempts` once but it turns out to be hugely useful on Redis SaaS platforms where Redis migrates or moves under the covers and Sidekiq is just supposed to recover gracefully.

**Comment by casperisfine** [comment]
> Do you see this driver as stable and feature complete by then?

I believe so yes, at least for Sidekiq use case (i.e. no redis-cluster).

Also not too sure about the `hiredis` driver, native extensions can be tricky to test properly, but I don't think it makes that big a difference for Sidekiq, so it might be marked as "experimental" in the couple first releases.

> I'm happy to adjust or remove some feature usage.

It's fine, `reconnect_attempts` is indeed very useful. I just want to implement it better (only retry idempotent commands). That's what I was working on this afternoon: https://github.com/redis-rb/redis-client/pull/9, it will probably be ready tomorrow-ish.

As for Sentinel, I need to have a look, but it shouldn't be too hard.

**Comment by casperisfine** [comment]
> reconnect_attempts, Sidekiq sets it to 1. However if we are to implement this, we should have COMMAND knowledge to only retry readonly tagged commands.

So I looked into having a list of commands safe to retry, but it's actually not that simple, because in many cases it also depends on the arguments and wether you care about the return value.

For example `ZREM ...` can be retried, but if the caller cares about the return value, then it might be incorrect to retry it.

So instead I added a `#call_once` method that prevent retries. So it basically allow to choose between AtMostOnce and AtLeastOnce consistency. I'll try to go over https://github.com/mperham/sidekiq/pull/5253 to see if there's places that shouldn't be retried.
